### ISSUE TITLE
ZLIB as shared library

### ISSUE THREAD

**Issue by isaachier** [ruslo/hunter#1103]
I am trying to package grpc (#561) and I am running into one major issue I can't figure out. Basically, grpc relies on ZLIB being built as a shared library, but only ZLIB. This has the drawback that I cannot use `HUNTER_ENABLE_SHARED_LIBS` because it would affect all of the dependencies. I even tried passing `ZLIB_BUILD_SHARED` in a custom `config.cmake`, but found this does not work when building the example for grpc against the grpc library using `jenkins.py`. My guess in that case is that the grpc config won't override the example project config, but not 100% sure.

TL;DR How can I build only ZLIB as a shared library while leaving everything else the way it is?

**Comment by ruslo** [comment]
Use `BUILD_SHARED_LIBS` in custom `CMAKE_ARGS`:
* https://github.com/ruslo/hunter/wiki/example.custom.config.id#custom-cmake-options

**Comment by isaachier** [comment]
This amounts to the same thing as `ZLIB_BUILD_SHARED` according to https://github.com/hunter-packages/zlib/blob/hunter/CMakeLists.txt#L186. That didn't work for me.

**Comment by isaachier** [comment]
It works to build grpc as a top-level project, but it fails when using `jenkins.py` script to build `examples/grpc`.

**Comment by ruslo** [comment]
How you're loading custom `config.cmake`?

Here is an example:
* https://github.com/ruslo/hunter/blob/b59f121e457bef2d20cc575653b668e194b3ca33/examples/OpenCV-extra/CMakeLists.txt#L4-L12
* https://github.com/ruslo/hunter/blob/b59f121e457bef2d20cc575653b668e194b3ca33/examples/OpenCV-extra/config.cmake#L1-L9

**Comment by isaachier** [comment]
Wait so that is my question/issue. Does this require every user of grpc to install the same custom config? Or can grpc have the custom config and have the main project automatically get that config?

**Comment by isaachier** [comment]
To be more specific, in the example above, OpenCV-extra is configuring OpenCV to build differently. Here is my case:

`examples/grpc` => `grpc` => `zlib` (shared object)

Is there any way for `grpc` to specify that `zlib` must be shared, or does every project have to ensure this (in this case `examples/grpc`)?

**Comment by ruslo** [comment]
> Does this require every user of grpc to install the same custom config?

Yes.

> Or can grpc have the custom config and have the main project automatically get that config?

You can set default `CMAKE_ARGS` for the package here:
* https://docs.hunter.sh/en/latest/creating-new/create/cmake.html#cmake-options

But I guess it's not what you want because we are not tuning `grpc`, you want to tune `zlib`. I can't set `BUILD_SHARED_LIBS=ON` for `zlib` since user expect `OFF` by default.

> does every project have to ensure this (in this case examples/grpc)?

Every project. I think the real problem is about that `grpc` is not working with static `zlib`. This one should be fixed.

**Comment by isaachier** [comment]
Right. They are trying to build some sort of C# extension from shared ZLIB. Minor extension, but major build implications.

**Comment by isaachier** [comment]
Actually, I need not make ZLIB shared, it just have to compile position independent code. Is this simpler to implement?

I guess I am a bit confused overall as to the status of unique configurations of dependencies. Does the config only matter at the top level? How can libraries set or even notify the user of the versions it needs for its dependencies?

For example:

`ProjectA` -> `LibraryB` -> `LibraryC`

Can `LibraryB` define anything about `LibraryC` without burdening `ProjectA` to add a custom config? If not, is there a user-friendly was for `LibraryB` to notify the owner of `ProjectA` to create a new `cmake/Hunter/config.cmake` file?

**Comment by ruslo** [comment]
> Is this simpler to implement?

Use toolchain, e.g.:
* https://github.com/ruslo/polly/blob/master/gcc-pic.cmake

> Does the config only matter at the top level?

Config used only once, either global or local. Second `HunterGate` command will have no effect and inherit all the properties of the very first `HunterGate`.

> How can libraries set or even notify the user of the versions it needs for its dependencies?

At this point only in documentation if you can't use defaults. In future it's possible to do something more convenient like https://github.com/ruslo/hunter/issues/1106.

> Can LibraryB define anything about LibraryC without burdening ProjectA to add a custom config?

And what should we do if `ProjectA` requirements conflicts with `LibraryB` requirements? At this point there is no such mechanism, custom `config.cmake` processed only once in top level and applied to all packages.

**Comment by isaachier** [comment]
I found a workaround for now. Thanks for the help.
